### PR TITLE
[SPARK-14849][CORE]Always set an address for the executor

### PR DESCRIPTION
**Pull request by skyluc**
As specified in [SPARK-14849](https://issues.apache.org/jira/browse/SPARK-14849), the `address` in the `NettyRpcEnv` for the executor is not set when running in standalone mode. To compensate, the driver try to guess the IP address of the executor during registration. But in a NAT situation, the IP address visible in the connection is not the one that should be used.
This address is sent back to the executor, and used later to describe the location of blocks.

The change is too always set the address as the `host` value, and to remove the code on the driver side which is trying to guess the right IP address. This kind of guessing is wrong in a NAT configuration, a possibly others.

Manually tested on the configuration describe on the ticket, and on a standard Mesos cluster.
Testing on a Yarn cluster would be useful, to check that there is no unexpected effects.

### CODE REVIEW

**Comment by rxin** [comment]
cc @zsxwing

**Comment by skyluc** [comment]
Are the test failures real, or due to flaky tests?
I tried to reproduced the failures locally, but `core/test` passes most of the time.

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala [hunk 1]
```diff
@@ -122,7 +122,7 @@ private[netty] class NettyRpcEnv(
 
   @Nullable
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala]: this is no longer nullable

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala [hunk 2]
```diff
@@ -122,7 +122,7 @@ private[netty] class NettyRpcEnv(
...
   override lazy val address: RpcAddress = {
-    if (server != null) RpcAddress(host, server.getPort()) else null
+    if (server != null) RpcAddress(host, server.getPort()) else RpcAddress(host, -1)
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala]: while you are at it, we should probably document when server is null and explain the choices here

**Comment by rxin** [comment]
I think there are also some code in SparkEnv that deals with this?
```
    if (isDriver) {
      conf.set("spark.driver.port", rpcEnv.address.port.toString)
    } else if (rpcEnv.address != null) {
      conf.set("spark.executor.port", rpcEnv.address.port.toString)
    }
```

**Comment by skyluc** [comment]
Added comments, updated checks for cases when running in client mode, removed sending back the hostname to the executor.

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala [hunk 3]
```diff
@@ -120,9 +120,15 @@ private[netty] class NettyRpcEnv(
...
   override lazy val address: RpcAddress = {
-    if (server != null) RpcAddress(host, server.getPort()) else null
+    // When running in client mode (i.e.: not opening a listening port, but connecting to a
+    // 'server' with an open port), the server value is null.
+    // But a valid host still has to be provided, as it is needed later on.
+    // Previously, the server side of the connection was attempting to guess the client
+    // hostname, using the connection information. But the value generated is wrong when
+    // the connection is NATed.
+    // [SPARK-14849]
+    if (server != null) RpcAddress(host, server.getPort()) else RpcAddress(host, -1)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala]: `-1` is actually confusing. All of executors in the same node will have the same address.

**Comment by zsxwing** [comment]
The host is easy to fix. However, the hard part is `port`. The executor cannot know the port until it sends a message. In addition, since the executor runs in the client mode, it may have multiple clients/ports.

**Comment by zsxwing** [comment]
@skyluc could you close this one, please? You can submit a new PR when you have a better idea. Thanks!

**Pull request closed without merging**